client/utils/useManyPubs/useManyPubs.ts

```diff
@@ -63,80 +67,92 @@ export const useManyPubs = <P extends Pub = Pub>(
+		return manyPubsState.current[queryKey] || initialQueryState;
+	}, [queryKey, manyPubsState]);
 
-	const setState = useCallback(
+	const setCurrentQueryState = useCallback(
 		(next: QueryState) => {
-			setManyPubsState((currentState) => {
-				return {
-					...currentState,
-					[queryKey]: next,
-				};
-			});
+			manyPubsState.current = {
+				...manyPubsState.current,
+				[queryKey]: next,
+			};
+			update();
 		},
-		[queryKey],
+		[queryKey, update, manyPubsState],
 	);
 
 	const loadMorePubs = async () => {
-		if (state.isLoading) {
+		const queryState = getCurrentQueryState();
+		const { isLoading, hasLoadedAllPubs, offset } = queryState;
+
+		if (isLoading || hasLoadedAllPubs) {
 			return;
 		}
 
-		const nextState = getStartLoadingPubsState(state, batchSize);
-		setState(nextState);
-
 		const query: ManyPubsQuery = {
 			...keyQuery,
 			...optionsQuery,
 			limit: batchSize,
-			offset: state.offset,
+			offset,
 			communityId: communityData.id,
 		};
+
+		const nextQueryState = getStartLoadingPubsState(queryState, batchSize);
+		setCurrentQueryState(nextQueryState);
+
 		const result: ManyPubsApiResult = await apiFetch.post('/api/pubs/many', {
 			query,
 			pubOptions,
-			alreadyFetchedPubIds: Object.keys(pubsById),
+			alreadyFetchedPubIds: Object.keys(pubsById.current),
 		});
 
 		const { loadedAllPubs, pubsById: newPubsById, pubIds } = result;
-		const nextPubsById = { ...pubsById, ...newPubsById };
+		const nextPubsById = { ...pubsById.current, ...newPubsById };
 		const resolvedPubs = pubIds.map((id) => nextPubsById[id]);
 		const resolvedPubsById = indexByProperty(resolvedPubs, 'id');
-		setState(getFinishedLoadingPubsState(nextState, query, resolvedPubsById, loadedAllPubs));
-		setPubsById(nextPubsById);
+		pubsById.current = nextPubsById;
+		setCurrentQueryState(
+			getFinishedLoadingPubsState(nextQueryState, query, resolvedPubsById, loadedAllPubs),
+		);
 	};
 
+	useUpdateEffect(() => {
+		if (!cacheQueries) {
+			pubsById.current = {};
+			manyPubsState.current = { [queryKey]: initialQueryState };
+		}
+	}, [cacheQueries, queryKey]);
+
 	useEffect(() => {
-		if (isEager && state.offset === 0) {
+		const queryState = getCurrentQueryState();
+		if (isEager && queryState.offset === 0) {
 			loadMorePubs();
 		}
 		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [isEager, queryKey, state.offset]);
+	}, [isEager, queryKey]);
 
+	const currentQueryState = getCurrentQueryState();
 	return {
 		currentQuery: {
 			loadMorePubs,
-			isLoading: state.isLoading,
-			hasLoadedAllPubs: state.hasLoadedAllPubs,
-			pubs: state.orderedPubs.pubsInOrder as P[],
+			isLoading: currentQueryState.isLoading,
+			hasLoadedAllPubs: currentQueryState.hasLoadedAllPubs,
+			pubs: currentQueryState.orderedPubs.pubsInOrder as P[],
 		},
 		allQueries: {
-			isLoading: Object.values(manyPubsState).some((s) => s.isLoading),
+			isLoading: Object.values(manyPubsState.current).some((s) => s.isLoading),
 		},
 	};
 };
```

Comment:
I like these changes. @howebenhowe mentioned implementing something like this a while back and you pulled it off great!

Reply:
It is kind of scary that the custom hooks I write all eventually get complicated enough that I need to use refs as an escape hatch to avoid the stale-closure problem. Oh well!